src/shared/sleep-config.c

```diff
@@ -33,18 +33,55 @@ static const char* const sleep_operation_table[_SLEEP_OPERATION_MAX] = {
 
 DEFINE_STRING_TABLE_LOOKUP(sleep_operation, SleepOperation);
 
+static char* const* const sleep_default_state_table[_SLEEP_OPERATION_CONFIG_MAX] = {
+        [SLEEP_SUSPEND]      = STRV_MAKE("mem", "standby", "freeze"),
+        [SLEEP_HIBERNATE]    = STRV_MAKE("disk"),
+        [SLEEP_HYBRID_SLEEP] = STRV_MAKE("disk"),
+};
+
+static char* const* const sleep_default_mode_table[_SLEEP_OPERATION_CONFIG_MAX] = {
+        /* Not used by SLEEP_SUSPEND */
+        [SLEEP_HIBERNATE]    = STRV_MAKE("platform", "shutdown"),
+        [SLEEP_HYBRID_SLEEP] = STRV_MAKE("suspend", "platform", "shutdown"),
+};
+
 SleepConfig* sleep_config_free(SleepConfig *sc) {
         if (!sc)
                 return NULL;
 
         for (SleepOperation i = 0; i < _SLEEP_OPERATION_CONFIG_MAX; i++) {
-                strv_free(sc->modes[i]);
                 strv_free(sc->states[i]);
+                strv_free(sc->modes[i]);
         }
 
         return mfree(sc);
 }
 
+static void sleep_config_validate_state_and_mode(SleepConfig *sc) {
+        assert(sc);
+
+        /* So we should really not allow setting SuspendState= to 'disk', which means hibernation. We have
+         * SLEEP_HIBERNATE for proper hibernation support, which includes checks for resume support (through
+         * EFI variable or resume= kernel command line option). It's simply not sensible to call the suspend
+         * operation but eventually do an unsafe hibernation. */
+        if (strv_contains(sc->states[SLEEP_SUSPEND], "disk")) {
+                strv_remove(sc->states[SLEEP_SUSPEND], "disk");
```

Comment:
frankly, this all looks as if this shouldn't be an strv in the first place. But a bit mask, i.e.:

```c
typedef enum SleepState {
        SLEEP_MEM, 
        SLEEP_STANDBY,
        SLEEP_FREEZE,
        SLEEP_DISK,
        _SLEEP_STATE_MAX,
        _SLEEP_STATE_INVALID = -EINVAL,
}; 
```

and similar for mode.

and then we should have tables somewhere:

```c
static const unsigned sleep_state_defaults[_SLEEP_OPERATION_MAX] = {
        [SLEEP_SUSPEND] = (1U << SLEEP_MEM) | (1U << SLEEP_STANDBY) | (1U << SLEEP_FREEZE),
        …
};

static const unsigned sleep_state_allowed[_SLEEP_OPERATION_MAX] = {
        [SLEEP_SUSPEND] = …
}
```

And then we default to the entry in the upper table, and only allow people to set the masks in the lower table. You get the idea.

Reply:
Hmm, actually we can't do that. Currently, we try the given states/modes in order. But if we switch to something like this it's no longer possible.